File: app/ios/OffTheCloud/OffTheCloud/PwCrypto.swift
import Foundation
import Security

// The bridge relays already-decrypted app payloads between a device and a
// friend/browser client, so a plaintext password in the protobuf envelope
// would be readable by the bridge operator. To keep the account password
// confidential end-to-end (see issue #2), the device generates an ephemeral
// RSA keypair per WebSocket connection and hands out the public half via
// GetPubKey/PubKey; the client encrypts the password with it
// (RSA-OAEP/SHA-256) before it ever leaves the app.

enum PwCryptoError: Error {
    case malformedPublicKey
    case importFailed(String)
    case encryptionUnsupported
    case encryptionFailed(String)
}

enum PwCrypto {
    /// Encrypts `password` with the PKIX/SPKI DER-encoded RSA public key
    /// returned by the device's GetPubKey/PubKey response.
    static func encryptPassword(_ password: String, pubKeyDER: Data) throws -> Data {
        let pkcs1 = try rsaPublicKeyPKCS1(fromSPKI: pubKeyDER)

        let attrs: [CFString: Any] = [
            kSecAttrKeyType: kSecAttrKeyTypeRSA,
            kSecAttrKeyClass: kSecAttrKeyClassPublic,
        ]
        var error: Unmanaged<CFError>?
        guard let secKey = SecKeyCreateWithData(pkcs1 as CFData, attrs as CFDictionary, &error) else {
            throw PwCryptoError.importFailed((error?.takeRetainedValue() as Error?)?.localizedDescription ?? "unknown error")
        }

        guard SecKeyIsAlgorithmSupported(secKey, .encrypt, .rsaEncryptionOAEPSHA256) else {
            throw PwCryptoError.encryptionUnsupported
        }

        var encError: Unmanaged<CFError>?
        guard let cipherText = SecKeyCreateEncryptedData(
            secKey,
            .rsaEncryptionOAEPSHA256,
            Data(password.utf8) as CFData,
            &encError
        ) as Data? else {
            throw PwCryptoError.encryptionFailed((encError?.takeRetainedValue() as Error?)?.localizedDescription ?? "unknown error")
        }

        return cipherText
    }

    /// The device sends the public key as X.509 SubjectPublicKeyInfo (what
    /// Go's x509.MarshalPKIXPublicKey produces), but SecKeyCreateWithData
    /// expects the bare PKCS#1 RSAPublicKey bytes. Strip the SPKI wrapper:
    /// SEQUENCE { AlgorithmIdentifier, BIT STRING { <PKCS#1 bytes> } }.
    private static func rsaPublicKeyPKCS1(fromSPKI der: Data) throws -> Data {
        var reader = DERReader(data: der)
        let outer = try reader.readElement(expectedTag: 0x30) // SEQUENCE
        var inner = DERReader(data: outer)
        _ = try inner.readElement(expectedTag: 0x30) // AlgorithmIdentifier, skip it
        let bitString = try inner.readElement(expectedTag: 0x03) // BIT STRING
        guard let unusedBits = bitString.first, unusedBits == 0 else {
            throw PwCryptoError.malformedPublicKey
        }
        return bitString.dropFirst()
    }
}

/// Minimal DER TLV reader, just enough to unwrap an SPKI structure.
private struct DERReader {
    let data: Data
    var index: Int = 0

    init(data: Data) { self.data = data }

    mutating func readElement(expectedTag: UInt8) throws -> Data {
        guard index < data.count, data[data.startIndex + index] == expectedTag else {
            throw PwCryptoError.malformedPublicKey
        }
        index += 1

        guard index < data.count else { throw PwCryptoError.malformedPublicKey }
        let firstLenByte = data[data.startIndex + index]
        index += 1

        let length: Int
        if firstLenByte & 0x80 == 0 {
            length = Int(firstLenByte)
        } else {
            let numLenBytes = Int(firstLenByte & 0x7F)
            guard numLenBytes > 0, numLenBytes <= 4, index + numLenBytes <= data.count else {
                throw PwCryptoError.malformedPublicKey
            }
            var len = 0
            for _ in 0..<numLenBytes {
                len = (len << 8) | Int(data[data.startIndex + index])
                index += 1
            }
            length = len
        }

        guard index + length <= data.count else { throw PwCryptoError.malformedPublicKey }
        let start = data.startIndex + index
        let content = data.subdata(in: start..<(start + length))
        index += length
        return content
    }
}

File: web/src/net/pwCrypto.ts
// SPDX-License-Identifier: AGPL-3.0-or-later

import forge from "node-forge";
import { ReqEnvelope, RespEnvelope } from "../proto/messages";

// The bridge relays already-decrypted app payloads between a device and a
// browser/friend client, so anything sent as plaintext in the protobuf
// envelope is readable by the bridge operator. To keep the account
// password confidential end-to-end (see issue #2), the device generates an
// ephemeral RSA keypair per WebSocket connection and hands out the public
// half via GetPubKey/PubKey; the client encrypts the password with it
// (RSA-OAEP/SHA-256) before it ever leaves the browser.
//
// This uses node-forge rather than the browser's native crypto.subtle:
// SubtleCrypto is only available in a "secure context" (HTTPS, or
// localhost/127.0.0.1), which a brand-new device being reached over plain
// http://<lan-ip-or-hostname> during first-run setup (issue #38/#39) isn't
// — crypto.subtle is simply undefined there, which used to crash this
// whole flow. node-forge does the same RSA-OAEP/SHA-256 math in pure JS,
// so it works regardless of secure-context status.

type Requester = (build: (e: Partial<ReqEnvelope>) => void) => Promise<RespEnvelope>;

// Issue #46: a plain browser tab has nothing else keeping the user signed
// in across a reload — the mobile app already gets the same effect for
// free by re-sending the Keychain-stored password on every launch (see
// App.tsx's `mobile` auto-auth effect). This mirrors that for the browser
// by persisting *something* across a reload and replaying it the same way
// on mount.
//
// Issue #101: that something used to be the actual account password,
// sitting in localStorage in plain text — anything that could read this
// origin's storage (an XSS, a shared/synced browser profile, a forensic
// grab of the profile directory) got the real password outright, not just
// a way back into one browser tab. It's now a random opaque token minted
// by the device itself (ReqIssueSessionToken/ReqAuthWithToken in
// messages.proto) that's good for nothing but resuming a session that was
// already established with the real password — see session/tokens.go for
// the server-side store this redeems against.
const cSessionTokenStorageKey = "otc_session_token";

// The token doesn't get to sit in localStorage forever just because the
// tab does — matches the old cPersistedKeyTTLMs window, and, like that
// one, keeps getting refreshed to a fresh hour on every successful
// redemption (see useWS.ts's authWithToken), so a tab reloaded regularly
// effectively never has to fall back to the password prompt.
const cPersistedTokenTTLMs = 60 * 60 * 1000;

interface PersistedToken {
  token: string;
  expiresAtMs: number;
}

export function savePersistedToken(token: string, expiresAtMs: number) {
  try {
    const entry: PersistedToken = { token, expiresAtMs };
    localStorage.setItem(cSessionTokenStorageKey, JSON.stringify(entry));
  } catch {
    // Storage can be unavailable (private browsing, quota) — session just
    // won't survive a reload in that case, not worth surfacing an error for.
  }
}

export function loadPersistedToken(): string | null {
  try {
    const raw = localStorage.getItem(cSessionTokenStorageKey);
    if (!raw) {
      return null;
    }
    const entry = JSON.parse(raw) as Partial<PersistedToken>;
    if (typeof entry.token !== "string" || typeof entry.expiresAtMs !== "number") {
      localStorage.removeItem(cSessionTokenStorageKey);
      return null;
    }
    // The server is the real authority on expiry (see session/tokens.go) —
    // this only skips a doomed round trip for a token that's already stale,
    // and distrusts an expiry claiming to sit further out than any token
    // the server would ever actually issue (a tampered or corrupted entry),
    // since a client-side check can only ever make this stricter, never
    // extend anything.
    if (Date.now() > entry.expiresAtMs || entry.expiresAtMs - Date.now() > cPersistedTokenTTLMs) {
      localStorage.removeItem(cSessionTokenStorageKey);
      return null;
    }
    return entry.token;
  } catch {
    return null;
  }
}

export function clearPersistedToken() {
  try {
    localStorage.removeItem(cSessionTokenStorageKey);
  } catch {
    // Nothing to clean up if storage isn't available in the first place.
  }
}

/**
 * Issue #39: every client calls GetPubKey before Auth anyway, and the
 * server rides along a `is_new_device` flag on that same response (true
 * when no owner secret has been set yet) — so this is a free way to tell
 * a fresh device apart from a normal login before showing a sign-in form.
 */
export async function isNewDevice(request: Requester): Promise<boolean> {
  const resp = await request((e) => {
    (e as any).payload = { $case: "reqGetPubKey", reqGetPubKey: {} };
  });

  if (resp.payload?.$case !== "respPubKey") {
    return false;
  }

  return resp.payload.respPubKey.isNewDevice;
}

/** Fetches this connection's public key and RSA-OAEP(SHA-256) encrypts `plaintext` with it. */
export async function encryptForConnection(request: Requester, plaintext: string): Promise<Uint8Array> {
  const resp = await request((e) => {
    (e as any).payload = { $case: "reqGetPubKey", reqGetPubKey: {} };
  });

  if (resp.payload?.$case !== "respPubKey") {
    const msg = resp.payload?.$case === "respAck" ? resp.payload.respAck.errorMsg : undefined;
    throw new Error(msg || "Unable to fetch the connection's public key");
  }

  const der = resp.payload.respPubKey.publicKey;
  // The server hands out PKIX/SubjectPublicKeyInfo DER (x509.MarshalPKIXPublicKey) —
  // exactly what forge.pki.publicKeyFromAsn1 expects.
  const derBuffer = forge.util.createBuffer(forge.util.binary.raw.encode(der));
  const publicKey = forge.pki.publicKeyFromAsn1(forge.asn1.fromDer(derBuffer)) as forge.pki.rsa.PublicKey;

  // Matches the server's rsa.DecryptOAEP(sha256.New(), ...): SHA-256 for
  // both the OAEP hash and MGF1, no label.
  const encrypted = publicKey.encrypt(
    forge.util.binary.raw.encode(new TextEncoder().encode(plaintext)),
    "RSA-OAEP",
    { md: forge.md.sha256.create(), mgf1: { md: forge.md.sha256.create() } },
  );

  return forge.util.binary.raw.decode(encrypted);
}
